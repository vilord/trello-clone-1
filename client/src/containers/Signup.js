import React, { Component } from 'react';
import { connect } from 'react-redux';
import { signupUser, loginUser } from '../actions/user';
import { Link } from 'react-router-dom';
import { Form, Button, Message } from 'semantic-ui-react';
import isEmail from 'validator/lib/isEmail';

import './Signup.css';
import gLogo from '../assets/G-logo.svg';

// TODO: Placeholder random fictional character API.
class Signup extends Component {
  constructor(props) {
    super(props);
    this.state = {
      name: '',
      email: '',
      password: '',
      emailError: false,
      passwordError: false,
      errorHeader: 'Header',
      errorMessage: 'Message',
    };

    // Bidings
    this.verifyEmail = this.verifyEmail.bind(this);
    this.handleChange = this.handleChange.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
  }

  componentWillReceiveProps(nextProps) {
    if (nextProps.userExists) {
      this.setState({
        errorHeader: 'User already exists!',
        errorMessage:
          'Email already in use by another account. You can use log in or use the forgot password page to reset your password.',
      });
    }
  }

  verifyEmail() {
    const { email } = this.state;
    this.setState({
      emailError: !isEmail(email),
      errorHeader: 'Verify your email!',
      errorMessage: "That doesn't look like an email address…",
    });
  }

  handleChange(e, { name, value }) {
    this.setState({ [name]: value });
  }

  handleSubmit(event) {
    event.preventDefault();

    const { name, email, password } = this.state;

    if (password.length < 8) {
      this.setState({
        passwordError: true,
        errorHeader: 'Password is to short!',
        errorMessage: 'Your password must be at least 8 characters.',
      });
      return;
    }

    const newUser = {
      name,
      email,
      password,
    };
    const { signupUser, history } = this.props;
    signupUser(newUser, history);
  }

  render() {
    const {
      name,
      email,
      password,
      emailError,
      passwordError,
      errorHeader,
      errorMessage,
    } = this.state;
    const { userExists } = this.props;

    const signupDisabled = !name || !email || !password || emailError;

    const host =
      process.env.NODE_ENV === 'development' ? 'http://localhost:3001' : '';
    const authURL = `${host}/auth/google`;

    return (
      <div className="Signup">
        <h1>Create an Account</h1>
        <p>
          or <Link to="/login">sign in to your account</Link>
        </p>
        <Form
          className="Signup-form"
          size="big"
          onSubmit={this.handleSubmit}
          error={emailError || passwordError || userExists}
        >
          <Form.Input
            autoFocus
            label="Name"
            type="text"
            name="name"
            value={name}
            placeholder="e.g., Hermione Granger"
            onChange={this.handleChange}
          />
          <Form.Input
            label="Email"
            type="email"
            name="email"
            value={email}
            placeholder="e.g., hermioneg@example.com"
            onBlur={this.verifyEmail}
            onChange={this.handleChange}
            error={emailError}
          />
          <Form.Input
            label="Password"
            type="password"
            name="password"
            value={password}
            placeholder="e.g., ••••••••••"
            onChange={this.handleChange}
            error={passwordError}
          />
          <Message error header={errorHeader} content={errorMessage} />
          <Button
            className="Signup-button"
            size="large"
            type="submit"
            fluid
            primary={!signupDisabled}
            disabled={signupDisabled}
          >
            Create New Account
          </Button>
        </Form>
        <a className="Signup-google" href={authURL}>
          <img src={gLogo} alt="" />
          <p>Sign up with Google</p>
        </a>
      </div>
    );
  }
}

const mapStateToProps = state => ({
  userExists: state.ui.userExists,
});

const mapDispatchToProps = dispatch => ({
  signupUser: newUser => {
    dispatch(signupUser(newUser));
  },
  loginUser: user => {
    dispatch(loginUser(user));
  },
});

export default connect(mapStateToProps, mapDispatchToProps)(Signup);
