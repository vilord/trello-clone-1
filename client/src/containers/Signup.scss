.Signup {
  display: flex;
  flex-direction: column;

  width: 350px;

  h1 {
    font-size: 3em;

    margin-bottom: 0;
  }

  p {
    font-size: 16px;
  }
}

.Signup-form {
  margin-top: 20px;
}

.Signup-google {
  font-family: 'Roboto', sans-serif;

  display: flex;
  align-items: center;
  justify-content: center;

  width: 100%;
  height: 46px;
  margin-top: 20px;

  color: grey;
  border: 2px solid lightgrey;
  border-radius: 5px;

  img {
    height: 70%;
    margin-right: 20px;
  }

  &:hover {
    color: grey;
  }

  &:active {
    background-color: lightgray;
  }
}
