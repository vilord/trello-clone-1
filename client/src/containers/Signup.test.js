import React from 'react';
import { shallow } from 'enzyme';
import { Signup } from './Signup';
import { Form } from 'semantic-ui-react';

describe('Signup Component', () => {
  beforeAll(() => {
    console.error = error => {
      throw new Error(error);
    };
  });

  const minProps = {
    setEmailError: jest.fn(),
    setPasswordError: jest.fn(),
    resetError: jest.fn(),
    signupUser: jest.fn(),
  };

  const state = {
    name: 'John Doe',
    email: 'johndoe@example.com',
    password: '1234abcd',
  };

  let component;
  beforeEach(() => {
    component = shallow(<Signup {...minProps} />);
  });

  it('renders', () => {
    expect(component).toHaveLength(1);
  });

  it('matches snapshot', () => {
    expect(component).toMatchSnapshot();
  });

  it('sets the values of inputs based on state (controlled component)', () => {
    component.find(Form.Input).forEach(x => {
      expect(x.prop('value')).toBe('');
    });

    component.setState(state);
    component.find(Form.Input).forEach(x => {
      expect(x.prop('value')).toBe(component.state(x.prop('name')));
    });
    expect(component).toMatchSnapshot();
  });

  it('sets error on email field on emailError', () => {
    expect(
      component
        .find(Form.Input)
        .at(1)
        .prop('error'),
    ).toBe(false);

    component.setProps({ emailError: true });

    expect(
      component
        .find(Form.Input)
        .at(1)
        .prop('error'),
    ).toBe(true);
  });

  it('sets error on password field on passwordError', () => {
    expect(
      component
        .find(Form.Input)
        .last()
        .prop('error'),
    ).toBe(false);

    component.setProps({ passwordError: true });

    expect(
      component
        .find(Form.Input)
        .last()
        .prop('error'),
    ).toBe(true);
  });

  it('enables button when all input fields have values', () => {
    expect(component.find('Button').prop('disabled')).toBe(true);
    component.setState(state);
    expect(component.find('Button').prop('disabled')).toBe(false);
  });

  it('disables button when sending', () => {
    component.setState(state);
    expect(component.find('Button').prop('disabled')).toBe(false);
    component.setProps({ sending: true });
    expect(component.find('Button').prop('disabled')).toBe(true);
  });
});
