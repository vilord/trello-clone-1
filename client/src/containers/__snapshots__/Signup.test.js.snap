// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Signup Component matches snapshot 1`] = `
<div
  className="Signup"
>
  <h1>
    Create an Account
  </h1>
  <p>
    or 
    <Link
      replace={false}
      to="/login"
    >
      sign in to your account
    </Link>
  </p>
  <Form
    as="form"
    className="Signup-form"
    error={false}
    onSubmit={[Function]}
    size="big"
  >
    <FormInput
      as={[Function]}
      autoFocus={true}
      control={[Function]}
      disabled={false}
      label="Name"
      name="name"
      onChange={[Function]}
      placeholder="e.g., Hermione Granger"
      type="text"
      value=""
    />
    <FormInput
      as={[Function]}
      control={[Function]}
      disabled={false}
      error={false}
      label="Email"
      name="email"
      onBlur={[Function]}
      onChange={[Function]}
      placeholder="e.g., hermioneg@example.com"
      type="email"
      value=""
    />
    <FormInput
      as={[Function]}
      control={[Function]}
      disabled={false}
      error={false}
      label="Password"
      name="password"
      onChange={[Function]}
      placeholder="e.g., ••••••••••"
      type="password"
      value=""
    />
    <Message
      content="Message"
      error={true}
      header="Header"
    />
    <Button
      as="button"
      className="Signup-button"
      disabled={true}
      fluid={true}
      primary={false}
      size="large"
      type="submit"
    >
      Create New Account
    </Button>
  </Form>
  <a
    className="Signup-google"
    href="/auth/google"
  >
    <img
      alt=""
      src="G-logo.svg"
    />
    <p>
      Sign up with Google
    </p>
  </a>
</div>
`;

exports[`Signup Component sets the values of inputs based on state (controlled component) 1`] = `
<div
  className="Signup"
>
  <h1>
    Create an Account
  </h1>
  <p>
    or 
    <Link
      replace={false}
      to="/login"
    >
      sign in to your account
    </Link>
  </p>
  <Form
    as="form"
    className="Signup-form"
    error={false}
    onSubmit={[Function]}
    size="big"
  >
    <FormInput
      as={[Function]}
      autoFocus={true}
      control={[Function]}
      disabled={false}
      label="Name"
      name="name"
      onChange={[Function]}
      placeholder="e.g., Hermione Granger"
      type="text"
      value="John Doe"
    />
    <FormInput
      as={[Function]}
      control={[Function]}
      disabled={false}
      error={false}
      label="Email"
      name="email"
      onBlur={[Function]}
      onChange={[Function]}
      placeholder="e.g., hermioneg@example.com"
      type="email"
      value="johndoe@example.com"
    />
    <FormInput
      as={[Function]}
      control={[Function]}
      disabled={false}
      error={false}
      label="Password"
      name="password"
      onChange={[Function]}
      placeholder="e.g., ••••••••••"
      type="password"
      value="1234abcd"
    />
    <Message
      content="Message"
      error={true}
      header="Header"
    />
    <Button
      as="button"
      className="Signup-button"
      disabled={false}
      fluid={true}
      primary={true}
      size="large"
      type="submit"
    >
      Create New Account
    </Button>
  </Form>
  <a
    className="Signup-google"
    href="/auth/google"
  >
    <img
      alt=""
      src="G-logo.svg"
    />
    <p>
      Sign up with Google
    </p>
  </a>
</div>
`;
